test/atomics.jl

```diff
@@ -381,3 +381,40 @@ ans = getfield(ARefxy{Any}(42, 42), :x, :sequentially_consistent, true)
 @test ans == 42
 ans = getfield(ARefxy{Any}(42, 42), :x, :sequentially_consistent, false)
 @test ans == 42
+
+struct StaticRef{value} end
+StaticRef(value) = StaticRef{value}()
+Base.getindex(::StaticRef{value}) where {value} = value
+
+@noinline do_modifyfield!(obj, field, op, x, order) =
+    modifyfield!(obj[], field[], op[], x[], order[])
+
+@testset "modifyfield!: partially inferrable arguments" begin
+    @testset for objw in [Ref, Ref{Any}],
+        fieldw in [StaticRef, Ref],
+        opw in [StaticRef, Ref],
+        xw in [Ref, Ref{Any}],
+        orderw in [StaticRef, Ref]
+
+        _modifyfield!(obj, field, op, x, order) =
+            do_modifyfield!(objw(obj), fieldw(field), opw(op), xw(x), orderw(order))
+
+        @test _modifyfield!(ARefxy(123, 0), :x, +, 456, :monotonic) ===
+              Pair{Int,Int}(123, 579)
+        @test _modifyfield!(ARefxy(123, 0), :x, Base.add_int, 456, :monotonic) ===
+              Pair{Int,Int}(123, 579)
+        @test _modifyfield!(ARefxy{Any}(123, 0), :x, +, 456, :monotonic) ===
+              Pair{Any,Any}(123, 579)
+        @test _modifyfield!(ARefxy{Any}(123, 0), :x, Base.add_int, 456, :monotonic) ===
+              Pair{Any,Any}(123, 579)
+        @test _modifyfield!(ARefxy(false, false), :x, |, true, :monotonic) ===
+              Pair{Bool,Bool}(false, true)
+        @test _modifyfield!(ARefxy(false, false), :x, Base.or_int, true, :monotonic) ===
+              Pair{Bool,Bool}(false, true)
+        @test_throws TypeError _modifyfield!(ARefxy(123, 0), :x, cglobal, 456, :monotonic)
+        @test_throws(
+            "`llvmcall` must be compiled to be called",
+            _modifyfield!(ARefxy(123, 0), :x, Core.Intrinsics.llvmcall, 456, :monotonic),
+        )
+    end
+end
```

Comment:
I think we'd want to have some tests on the level of Julia IR to assert the functionality of  `optimize_modifyop!`. Something like [these test cases](https://github.com/JuliaLang/julia/blob/ef1c9ce263390d23f9419345fe5902edec344433/test/compiler/inline.jl#L751-L765) might be useful for implementing it (there is also the `test/compiler/irutils.jl` utility script).

Reply:
This is actually more about testing that the codegen can handle edge cases with `modifyfield!` and handwritten intrinsic functions. But I agree it's a good idea to directly test that the optimization on IRCode is working as expected.

Reply:
Implemented in f92dead06474b7f4444b04793e749799dbc77b42